add script to create secrets for Prometheus and Grafana (#883)
redhat-appstudio/infra-deployments: README.md

#### Workspaces:
If `-rw | --root-workspace` parameter is not specified, then by default, all workspaces are automatically created under the `root` workspace.
There are two workspaces created per kcp instance:

> * `redhat-appstudio-internal-compute` - This is the workspace where the SyncTarget for the OpenShift Prometheus cluster is configured. If the root workspace is different from `root`, then the name of the workspace is set to `compute` to work around [this issue](https://github.com/kcp-dev/kcp/issues/1843). (The name of the workspace can be overridden by setting the `COMPUTE_WORKSPACE` variable)

Author: reverted in dd05d11

To configure the kcp instances separately use the `configure-kcp.sh` script:
```bash
./hack/configure-kcp.sh -kk [kubeconfig-pointing-to-kcp] -ck [kubeconfig-pointing-to-openshift] -rw [workspace-to-be-used-as-root] -kn [kcp-name|kcp-stable,kcp-unstable,dev]
```
which takes care of creation of the workspaces, SyncTarget, and the representation of the cluster in ArgoCD for the given kcp instance. So to fully finish the upstream configuration run the script for both kcp instances using the parameter `-kn kcp-stable` and `-kn kcp-unstable`.

#### Access Argo CD Web UI
Open the Argo CD Web UI to see the status of your deployments. You can use the route from the previous step and login using your OpenShift credentials (using the 'Login with OpenShift' button), or login to the OpenShift Console and navigate to Argo CD using the OpenShift Gitops menu in the Applications pulldown.
   ![OpenShift Gitops menu with Cluster Argo CD menu option](documentation/images/argo-cd-login.png?raw=true "OpenShift Gitops menu")

If your deployment was successful, you should see several applications running, such as "all-components", "application-service", and so on.

> ### Optional: OpenShift Local (formerly CRC) Post-Bootstrap Configuration

Reviewer: I guess that this is a zombie :ghost: from the kcp/pre-kcp branch switch.

Author: removed in 67ce1d0

For access to the OpenShift staging cluster, the user must be added to the `stage` team in the `redhat-appstudio-sre` Github organization.

## Monitoring for workload clusters

Both Prometheus and Grafana UIs are protected by an OAuth2 proxy which delegates the authentication to GitHub. 

> Users must belong to the [Red Hat Appstudio SRE organization](https://github.com/redhat-appstudio-sre) team configured in the OAuth2 proxy to be allowed to access these UIs.

Author: AFAIK, yes, as long as you have the values (or you have access to the settings to generate a new secret).

> Users must belong to the [Red Hat Appstudio SRE organization](https://github.com/redhat-appstudio-sre) team configured in the OAuth2 proxy to be allowed to access these UIs.

Author: @Michkov I stand corrected: you'll need to configure your OAuth2 proxy with the client id and secret, and GitHub will redirect to the callback URL configured in the settings, so basically a single cluster at a time that can use a given client id/secret pair.

> Users must belong to the [Red Hat Appstudio SRE organization](https://github.com/redhat-appstudio-sre) team configured in the OAuth2 proxy to be allowed to access these UIs.

Reviewer 2: Just thinking if it should not be described in generic way and then subsection about configuration of staging instance. (but this is not blocking merge of this PR from my side)

> Users must belong to the [Red Hat Appstudio SRE organization](https://github.com/redhat-appstudio-sre) team configured in the OAuth2 proxy to be allowed to access these UIs.

Reviewer 3: I think this is explained more clearly now.

### Setup

Make sure you have the following environment variables:
- `PROMETHEUS_GITHUB_CLIENT_ID`
- `PROMETHEUS_GITHUB_CLIENT_SECRET`
- `PROMETHEUS_GITHUB_COOKIE_SECRET`
- `GRAFANA_GITHUB_CLIENT_ID`
- `GRAFANA_GITHUB_CLIENT_SECRET`
- `GRAFANA_GITHUB_COOKIE_SECRET`

> The `PROMETHEUS_GITHUB_CLIENT_ID`/`PROMETHEUS_GITHUB_CLIENT_SECRET` and `GRAFANA_GITHUB_CLIENT_ID`/`GRAFANA_GITHUB_CLIENT_SECRET` value pairs must match an existing OAuth application on GitHub - see [OAuth apps](https://github.com/organizations/redhat-appstudio-sre/settings/applications) in the [Red Hat Appstudio SRE organization](https://github.com/organizations/redhat-appstudio-sre). The `PROMETHEUS_GITHUB_COOKIE_SECRET` and `GRAFANA_GITHUB_COOKIE_SECRET` can be generated using the [followiing instructions](https://oauth2-proxy.github.io/oauth2-proxy/docs/configuration/overview#generating-a-cookie-secret).

Author: thanks!

For access to the OpenShift staging cluster, the user must be added to the `stage` team in the `redhat-appstudio-sre` Github organization.

## Monitoring for workload clusters

Note:

This section refers to "Grafana cluster" and "Prometheus cluster" as the clusters on which Grafana and Prometheus are deployed, respectively. In a multi-cluster topology, there will be a single cluster on which Grafana is deployed, whereas Prometheus will be deployed on all clusters where collecting metrics is needed.

### Setup

> First, create the `appstudio-workload-monitoring` namespace if it does not exist yet:

Reviewer 3: ```suggestion
First, create the `appstudio-workload-monitoring` namespace on each Prometheus or Grafana cluster, if it does not exist yet:
```

```
$ oc create namespace appstudio-workload-monitoring

> ```

Reviewer 3: Do you want them to run this command, or
```
$ ./hack/setup-monitoring.sh  namespace
```

Author: I initially had a function for this in the script, but then I thought it was maybe not needed, for such a simple command. However, I can restore it if it makes things easier. Also, IIRC @bamachrn said that the namespace would be created by ArgoCD? (I haven't verified that yet, though)

Author: would you have any pointers to this, please?

Reviewer 3: It's actually automatically creating the namespaces when installing the operators (using the CreateNamespace=true directive), rather than explicitly creating the namespaces.  Here's an example: https://github.com/redhat-appstudio/infra-deployments/blob/4f7edc301a7d01d7710cb75aff7b2c41a1c203f1/argo-cd-apps/base/has.yaml#L28

Reviewer 3: You're right that it's a simple command, but if we can automate creating the namespace, even better.  I just don't have sample code handy for that.  I wouldn't put a ton of effort into automating this one line, since we're already doing this semi-manually.

Author: ok, I suggest we address that in a separate PR, once we have a `./components/monitoring/kustomize.yaml` in place (it's currently missing after the `main` branch was switch to pre-KCP)

#### OAuth2 proxy secrets

Both Prometheus and Grafana UIs are protected by an OAuth2 proxy running as a sidecar container and which delegates the authentication to GitHub. 
Users must belong to the [Red Hat Appstudio SRE organization](https://github.com/redhat-appstudio-sre) team configured in the OAuth2 proxy to be allowed to access the Web UIs.

Create the secrets with the following commands:

```
$ ./hack/setup-monitoring.sh oauth2-secret prometheus-oauth2-proxy $PROMETHEUS_GITHUB_CLIENT_ID $PROMETHEUS_GITHUB_CLIENT_SECRET $PROMETHEUS_GITHUB_COOKIE_SECRET

$ ./hack/setup-monitoring.sh oauth2-secret grafana-oauth2-proxy $GRAFANA_GITHUB_CLIENT_ID $GRAFANA_GITHUB_CLIENT_SECRET $GRAFANA_GITHUB_COOKIE_SECRET

> ```

Reviewer 3: I think you should specify here that the promethus-oauth2-proxy secret needs to be created for each Prometheus cluster.  And do you need a different "OAuth application" for each cluster or is it the same one for all clusters?

And then specify that you only need to create the grafana-oauth2-proxy secret once, on the Grafana cluster.

Author: good point, let me clarify that! And yes, it's a different OAuth Application for each cluster, indeed

Author: done in 13337c0

Reviewer 4: this means we need to update the env variables before running for each prometheus cluster. could we put this somewhere?

Author: > could we put this somewhere?

did you mean documenting it? If so, please see in 98c9d0f

Reviewer 3: These changes look good to me, thanks.

The `PROMETHEUS_GITHUB_CLIENT_ID`/`PROMETHEUS_GITHUB_CLIENT_SECRET` and `GRAFANA_GITHUB_CLIENT_ID`/`GRAFANA_GITHUB_CLIENT_SECRET` value pairs must match an existing "OAuth application" on GitHub - see [OAuth apps](https://github.com/organizations/redhat-appstudio-sre/settings/applications) in the [Red Hat Appstudio SRE organization](https://github.com/organizations/redhat-appstudio-sre). The `PROMETHEUS_GITHUB_COOKIE_SECRET` and `GRAFANA_GITHUB_COOKIE_SECRET` can be generated using the [following instructions](https://oauth2-proxy.github.io/oauth2-proxy/docs/configuration/overview#generating-a-cookie-secret).

#### Grafana Datasources

> Grafana datasources contain the connection settings to the Prometheus instances. These datasources are stored in secrets in the `appstudio-workload-monitoring` of the **Grafana cluster**.

Reviewer 3: ```suggestion
Grafana datasources contain the connection settings to the Prometheus instances. These datasources are stored in secrets in the `appstudio-workload-monitoring` namespace of the **Grafana cluster**.
```

The Prometheus endpoints used by Grafana are protected by an OAuth proxy running as a sidecar container and which checks that the incoming requests contain a valid token. A token is valid if it belongs to a service account of the **Prometheus cluster** and which has the RBAC permission to "get namespaces". This can be obtained with the `cluster-monitoring-view` cluster role.

```
$ ./hack/setup-monitoring.sh grafana-datasource-secret $DATASOURCE_NAME $PROMETHEUS_URL $GRAFANA_TOKEN
```

> `DATASOURCE_NAME` is the name of the secret itself, as well as the base name of the YAML file it contains. Grafana supports datasources from multiple secrets, so their names must be unique.

Reviewer 3: I'm not sure what to provide for $DATASOURCE_NAME from this doc. What would be an example of a datasource name?  Does it have to correspond to some resource or namespace that already exists, or can it be an arbitrary string?

Author: clarified in 3f1a2da

Reviewer 3: Perfect, thanks.

`PROMETHEUS_URL` is obtained from the route created for Prometheus in the `openshift-monitoring` and `appstudio-workload-monitoring` namespaces:

```
$ PROMETHEUS_URL=`oc get route/prometheus-k8s -n openshift-monitoring -o json | jq -r '.status.ingress[0].host'`

$ PROMETHEUS_URL=`oc get route/prometheus -n appstudio-workload-monitoring -o json | jq -r '.status.ingress[0].host'`
```

`GRAFANA_TOKEN` is obtained by requesting a token for the `grafana-oauth` service account in the **Prometheus cluster**:
```
$ GRAFANA_TOKEN=`oc create token grafana-oauth -n appstudio-workload-monitoring`

> ```

Reviewer 3: Do we need to ask users to enter the $PROMETHEUS_URL and $GRAFANA_TOKEN or can the script calculate them automatically?

Author: In a multi-cluster setup, each pair of `$PROMETHEUS_URL` and `$GRAFANA_TOKEN` values will be retrieved from the workload cluster were Prometheus is deployed, but used in the "control plane" cluster where the Grafana datasource secret is going to be created. So for now I believe it's simpler to ask the user to do it manually.

Reviewer 3: OK that makes sense.  I would recommend restructuring this section in the order that people would run the commands, to make it easier to follow.

> #### Grafana Datasources
> 
> Grafana datasources contain the connection settings to the Prometheus instances. These datasources are stored in secrets in the `appstudio-workload-monitoring` namespace of the **Grafana cluster**.
> 
> The Prometheus endpoints used by Grafana are protected by an OAuth proxy running as a sidecar container and which checks that the incoming requests contain a valid token. A token is valid if it belongs to a service account of the **Prometheus cluster** and which has the RBAC permission to "get namespaces". This can be obtained with the `cluster-monitoring-view` cluster role.
>
> In a multi-cluster setup, each pair of $PROMETHEUS_URL and $GRAFANA_TOKEN values will be retrieved from the workload cluster were Prometheus is deployed, but used in the "control plane" cluster where the Grafana datasource secret is going to be created.  Run the following commands to obtain the tokens from each workload cluster where Prometheus is deployed:

(and then the commands to get those $PROMETHEUS_URL and $GRAFANA_TOKEN values)

> Using the values you just obtained from the **Prometheus cluster**, run the following command on the **Grafana cluster**:
> 

(and then the grafana-datasource-secret command and explanation about the arbitrary DATASOURCE_NAME that you recently added)

> Repeat this procedure for each workload cluster where Prometheus is deployed.

Author: I've updated that section in 01b893b, could you give it another look and let me know if it's ok?

Reviewer 3: Per our conversation in Slack, please add `kustomize build components/monitoring/base | oc apply -f - `

Reviewer 3: I see that in the latest commit, thanks.

> The `grafana-oauth` service account is created by `.components/monitoring/base/configure-prometheus.yaml` along with a binding to the `cluster-monitoring-view` cluster role. 

Reviewer 3: How do we get Argo CD to apply this configuration?

Author: I'm not sure I understand your question: this `grafana-oauth` service account and its associated cluster role binding are defined in `components/monitoring/base/configure-prometheus.yaml`, so it going to be managed by ArgoCD, isn't it?

Reviewer 3: I haven't gotten far enough to see if it works.  Since it's in the base configuration, I think this prometheus configuration will be applied to every cluster where you run the bootstrap script.  Let's verify that and then I would add that statement here to avoid confusion.  Otherwise people might think they need to `kubectl apply` it.

Author: ok, makes sense. Yes, it should be applied by ArgoCD on every cluster (as opposed to Grafana on a single cluster), but we should mention that for the sake of clarity. I'll add that once we have `./components/monitoring` in place.

Users can be added to organizations by Michal Kovarik <mkovarik@redhat.com> and by Shoubhik Bose <shbose@redhat.com>.

## Monitoring for Prometheus clusters

Note:

> This section refers to "Grafana cluster" and "Prometheus cluster" as the clusters on which Grafana and Prometheus are deployed, respectively. In a multi-cluster topology, there will be a single cluster on which Grafana is deployed, whereas Prometheus will be deployed on all clusters where metrics needed to be collected.

Reviewer 3: ```suggestion
This section uses **Grafana cluster** and **Prometheus cluster** to refer to the clusters on which Grafana and Prometheus are deployed, respectively. In a multi-cluster topology, there will be a single cluster on which Grafana is deployed, whereas Prometheus will be deployed on all clusters where metrics needed to be collected.
```

### Setup

First, create the `appstudio-workload-monitoring` namespace on each Prometheus or Grafana cluster, if it does not exist yet:

```
$ oc create namespace appstudio-workload-monitoring
```

#### OAuth2 proxy secrets

Both Prometheus and Grafana UIs are protected by an OAuth2 proxy running as a sidecar container and which delegates the authentication to GitHub. 
Users must belong to the [Red Hat Appstudio SRE organization](https://github.com/redhat-appstudio-sre) team configured in the OAuth2 proxy to be allowed to access the Web UIs.

Create the secrets with the following commands:

```
# on each Prometheus cluster
$ ./hack/setup-monitoring.sh oauth2-secret prometheus-oauth2-proxy $PROMETHEUS_GITHUB_CLIENT_ID $PROMETHEUS_GITHUB_CLIENT_SECRET $PROMETHEUS_GITHUB_COOKIE_SECRET

# on the Grafana cluster
$ ./hack/setup-monitoring.sh oauth2-secret grafana-oauth2-proxy $GRAFANA_GITHUB_CLIENT_ID $GRAFANA_GITHUB_CLIENT_SECRET $GRAFANA_GITHUB_COOKIE_SECRET
```

> The `PROMETHEUS_GITHUB_CLIENT_ID`/`PROMETHEUS_GITHUB_CLIENT_SECRET` and `GRAFANA_GITHUB_CLIENT_ID`/`GRAFANA_GITHUB_CLIENT_SECRET` value pairs must match an existing "OAuth Application" on GitHub - see [OAuth apps](https://github.com/organizations/redhat-appstudio-sre/settings/applications) in the [Red Hat Appstudio SRE organization](https://github.com/organizations/redhat-appstudio-sre). The `PROMETHEUS_GITHUB_COOKIE_SECRET` and `GRAFANA_GITHUB_COOKIE_SECRET` can be generated using the [following instructions](https://oauth2-proxy.github.io/oauth2-proxy/docs/configuration/overview#generating-a-cookie-secret).

Reviewer 3: Not a blocker though.

> The `PROMETHEUS_GITHUB_CLIENT_ID`/`PROMETHEUS_GITHUB_CLIENT_SECRET` and `GRAFANA_GITHUB_CLIENT_ID`/`GRAFANA_GITHUB_CLIENT_SECRET` value pairs must match an existing "OAuth Application" on GitHub - see [OAuth apps](https://github.com/organizations/redhat-appstudio-sre/settings/applications) in the [Red Hat Appstudio SRE organization](https://github.com/organizations/redhat-appstudio-sre). The `PROMETHEUS_GITHUB_COOKIE_SECRET` and `GRAFANA_GITHUB_COOKIE_SECRET` can be generated using the [following instructions](https://oauth2-proxy.github.io/oauth2-proxy/docs/configuration/overview#generating-a-cookie-secret).

Reviewer 3: Not a blocker though.

#### Grafana Datasources

Grafana datasources contain the connection settings to the Prometheus instances. These datasources are stored in secrets in the `appstudio-workload-monitoring` namespace of the **Grafana cluster**.

The Prometheus endpoints used by Grafana are protected by an OAuth proxy running as a sidecar container and which checks that the incoming requests contain a valid token. A token is valid if it belongs to a service account of the **Prometheus cluster** and which has the RBAC permission to "get namespaces". This can be obtained with the `cluster-monitoring-view` cluster role.

```
$ ./hack/setup-monitoring.sh grafana-datasource-secret $DATASOURCE_NAME $PROMETHEUS_URL $GRAFANA_TOKEN
```

`DATASOURCE_NAME` is the name of the secret itself, as well as the base name of the YAML file it contains. Grafana supports datasources from multiple secrets, so their names must be unique.
`DATASOURCE_NAME` is an arbitrary name, for example `cluster-1-prometheus-openshift-ds` for Prometheus running in the `openshift-monitoring` namespace of Cluster-1.

> `PROMETHEUS_URL` is obtained from the route created for Prometheus in the `openshift-monitoring` and `appstudio-workload-monitoring` namespaces:

Reviewer 3: ```suggestion
`PROMETHEUS_URL` is obtained from the route created for Prometheus in the `openshift-monitoring` and `appstudio-workload-monitoring` namespaces in the **Prometheus cluster**:
```

Reviewer 3: Because we're jumping between clusters, I added the **Prometheus cluster** here to make it easier even though you could infer it.

There are two commands below to get the PROMETHUS_URL, but only one place to use that URL.  Do these two commands return the same value (in which case we can just provide one way to get it)?  Or do I need to use both values somehow?

```
$ PROMETHEUS_URL=`oc get route/prometheus-k8s -n openshift-monitoring -o json | jq -r '.status.ingress[0].host'`

$ PROMETHEUS_URL=`oc get route/prometheus-oauth -n appstudio-workload-monitoring -o json | jq -r '.status.ingress[0].host'`
```

`GRAFANA_TOKEN` is obtained by requesting a token for the `grafana-oauth` service account in the **Prometheus cluster**:
```
$ GRAFANA_TOKEN=`oc create token grafana-oauth -n appstudio-workload-monitoring`
```

> The `grafana-oauth` service account is created by `components/monitoring/base/configure-prometheus.yaml` along with a binding to the `cluster-monitoring-view` cluster role. 

Reviewer 3: Please explain how configure-prometheus.yaml gets applied here. Do I need to run another command to apply this using Argo? Or is this already applied to all clusters if you run the bootstrap or preview scripts?
